README.txt
OSHA ESENER surveyanswers
*************************

.. contents::

.. Note!
   -----
   
   - code repository
   - bug tracker
   - questions/comments feedback mail


- Code repository: http://svn.syslab.com/svn/OSHA/osha.surveyanswers
- Questions and comments to info (at) syslab (dot) com
- Report bugs at http://products.syslab.com/products/osha.surveyanswers



Credits
=======

Copyright European Agency for Health and Safety at Work and Syslab.com
GmbH.

osha.surveyanswers development was funded by the European Agency for
Health and Safety at Work.


License
=======

osha.surveyanswers is licensed under the GNU Lesser Generic Public
License, version 2 or later and EUPL version 1.1 only. The complete
license texts can be found in docs/LICENSE.GPL and docs/LICENSE.EUPL.


CHANGES.txt
Changelog
=========

2.0.2 (unreleased)
------------------

- Nothing changed yet.


2.0.1 (2013-08-22)
------------------

- #8207 adjust CSS to make scrolling possible with long menu items
  [deroiste]


2.0 (2012-09-17)
----------------

- Plone 4

1.9.7 (2011-02-21)
------------------

- removing Products.OSHATranslations as dependeny, translations are now in osha.theme
  [thomasw] 

1.9.6 (2010-11-23)
------------------

- Fixed an issue with translated countries. Fixes #2213 (gerken)


1.9.5 (2010-06-13)
------------------

- 2 translation fixes #1321 (thomasw)


1.9.4 (2010-06-01)
------------------

- Added the translations for the missing strings, received from the Agency [thomasw]

1.9.3 (2010-05-26)
------------------

- Some more translations (Question-group labels) in 3 langs (thomasw) 

1.9.2 (2010-05-06)
------------------

- Created a missing_xx.po for every langusge (thomasw)
- Display ESENER logo in title (thomasw) 
- Wrote custom language selector that preserves the selected question (thomasw)
- insert horizontal scroll bar below chart if the browser window gets too narrow (thomasw)
- new colour scheme for the map (thomasw)
- Corrected botched translations, e.g. translating "No" with "Not Difficult At All" (thomasw)


1.9.1 (2010-03-23)
------------------

- Added EUPL license (deroiste)
- deactivated stylesfile, they are now handled in subsite.css in /sub/esener (conflicting statements removed) (pilz)
- added question number in question overview, order question according to that
  number (gerken)
- totally changed excel for question views (gerken)
- Resolved issue with flash overlapping menu (gerken)
- make sure the highlighting for the "discriminators" box works (thomasw)
- manually fixed tons of translations that have an awkward format in the database,
  such as trailing space or substituted M$ character (thomasw)
- Disabled the translation of country names, since they appear in vertical labels,
  and FusionCharts con only display ASCII for that (thomasw)
- Improved lay-out of map and floating discriminators box, less space used (thomasw)
- Added highlighting in menue based on currently selected question
- The labels on the single-country view are no longer vertical, since they might
  contain non-ASCII (thomasw)
- sort the countries in the chart by their abbreviation #932 (thomasw)


1.9.0 (2009-10-21)
------------------

- Changed JS, questions pop up differently now (gerken)
- Nice styling (pecek)

1.8.1 (2009-10-07)
------------------

- Fixed country mapping error (gerken)
- Removed two more questions from import (gerken)
- Countries that do not participate don't have two letter codes and are white
  now (gerken)
- Ensuring now that on charts the yAxis never goes above 100% (gerken)


1.8.0 (2009-10-02)
------------------

- New translation for question groups (gerken)
- Fixed an issue with the right layer of our skin (gerken)
- Changed algorithm for calculating percentage with grouped results (gerken)
- Added caching (gerken)
- Added prime script to prime the cache (gerken)

1.7.0 (2009-09-28)
------------------

- New import scripts for final data (gerken)

1.6.0 (2009-09-22)
------------------

- Added support for questions with too few answers (gerken)

1.5.2 (2009-09-21)
------------------

- We now ignore answers where no answer was given
  These questions were actually never asked (gerken)
- Now with the real meanings of categories (sec3, size_5) (gerken)
- Complete country mapping no answers. Norway and Switzerland added (gerken)
- Added support for % questions (gerken)

1.4.0 (2009-09-15)
------------------

- Added missing translation (gerken)
- Added fallback if no map default answer has no translation (gerken)
- Fixed multiple query string issues (gerken)
- Fixed a wrong calculation if noone from a given country answered (gerken)
- Updated all DB Scripts for usage with postgres. No compatibility with sqlite retained. Therefor a bigger release (gerken)

1.3.2 (2009-09-14)
------------------

- Make it work with postgres (gerken)

1.3.1 (2009-09-10)
------------------

- Slightly improved error handling (gerken)

1.3 (2009-09-10)
----------------

- Some improvements in translations (gerken)
- More stable adapters (gerken)

1.2 (2009-09-08)
----------------

- Much more working. Translations in part (gerken)

1.1 (2009-06-10)
----------------

- I screwed up the first release, here goes another one (thomasw)

1.0 (2009-06-10)
----------------

- First release (thomasw)

0.1 (2008-03-31)
----------------

- Initial creation



db_scripts/questions.py
import re
import random

country_mapping = {
12:3,
1:5,
31:7,
46:8,
32:44,
33:9,  
2:10,
34:11,
6:12,
7:13,
3:14,
4:15,
35:16,
8:18,
9:19,
36:20,
37:22,
10:23,
38:25,
11:29,
39:31,
13:32,
40:33,
41:36,
42:37,
5:38,
14:39,
43:45,
15:42,
51:40,
52:30,
}
percent_questions = ['mm401', 'mm405', 'mm400', 'rev1_1di']
questions = []
finder = re.compile('([^ ]+) +[^ ]+ +([^ ]+) +([^ ]+).*([01]+)')
for line in file('questions.txt'):
    data = list(finder.findall(line)[0])
    data[0] = data[0].replace(".", "")
    data[1] = int(data[1]) - 1
    data[2] = int(data[2])
    questions.append(data)

groups = ['group %i' % i for i in range(10)]

print ("delete from questions;")
for question in questions:
    options = dict(
        question_field = question[0],
        question = question[0],
        question_group = random.sample(groups, 1)[0],
        is_country = str(question[0] == 'country' and '1' or '0'),
        is_designator = str(question[0] in ('sec3', 'size_5') and '1' or '0'),
        type = question[0].lower() in percent_questions and '2' or '1',
        show_which = question[0].lower() in percent_questions and '0' or '2',
        show_which_text = 'yes', 
        hide = question[3])
    print 'insert into questions (question_field, question, question_group, is_country, is_designator, type, show_which, show_which_text, hide_question) values(\'%(question_field)s\', \'%(question)s\', \'%(question_group)s\', %(is_country)s, %(is_designator)s, %(type)s, %(show_which)s, \'%(show_which_text)s\', %(hide)s);' % options
    if options['is_country'] == '1':
        print 'update answer_meanings set question_id = (select id from questions where question_field = \'%s\') where question_id = 0;' % options['question_field']
        print 'delete from answer_meanings where question_id != (select id from questions where question_field = \'%s\');' % options['question_field']

print 'drop table responses;'
print 'create table responses (id SERIAL PRIMARY KEY, %s);' % (", ".join(["%s %s" % (x[0], x[0].lower() in percent_questions and 'FLOAT' or 'INTEGER') for x in questions[:-4]] + ["%s FLOAT" % x[0] for x in questions[-4:]]))
counter = 0
for line in file('ESENER2009_20090924_EUOSHA.dat'):
    sql_question = []
    sql_answer = []
    for question in questions:
        sql_question.append(question[0])
        answer = line[question[1]:question[2]].strip()
        if answer.startswith('.'):
            answer = '0' + answer
        answer = answer.replace(',', '')
        if answer == '':
            answer = 0
        if question[0] == 'country':
            answer = str(country_mapping[int(answer)])
        else:
            try:
                if question[0].lower() not in percent_questions + ['country2']:
                    answer = str(2**int(answer))
                else:
                    answer = '%02.2f' % float(answer)
            except:
                pass
        sql_answer.append(answer)
    print 'insert into responses(%s) values (%s);' % (", ".join(sql_question), ", ".join(['\'%s\'' % x for x in sql_answer]))
    
